### PR TITLE
ci/python-publish: bump, use trusted publishing

### LINKED ISSUE #2344: Update python-publish.yml to use Trusted Publishing
I'm filing this for myself to work on 🙂 

Per https://github.com/github/docs/pull/32146 and the earlier https://github.com/github/docs/pull/24772: PyPI now supports authentication via OIDC ("Trusted Publishing") and other official GitHub resources mostly refer to that authentication mechanism.

This workflow should be similarly updated to use Trusted Publishing instead of a manually configured API token.

### PR DESCRIPTION
**Pull request by woodruffw**
Fixes #2344.

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [x] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [x] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [x] Should specify least privileged [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully.

**For _CI_ workflows, the workflow:**

- [x] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [ ] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [x] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [x] Packaging workflows should run on `release` with `types: [ created ]`.
- [x] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**Some general notes:**

- [x] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [x] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [x] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [x] Automation and CI workflows cannot be dependent on a paid service or product.

### CODE REVIEW

**Comment by woodruffw** [comment]
xref https://github.com/actions/starter-workflows/issues/2344

CC @di @jhutchings1 

CC @webknjaz as well, as the maintainer of `gh-action-pypi-publish` 🙂

**Review by di — APPROVED** [review]

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:15]: It might be good to keep this globally in case people start modifying the jobs post workflow creation.

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:26]:
```suggestion
      - name: Build release distributions
```
(same)

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:30]: *(this is probably controversial but)* the CI images come with pipx pre-provisioned so this could just use that:
```suggestion
          pipx run --spec build python -Im build
```

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:42]: It'd *really* want to get people to use environment-based protection from the very beginning:
```suggestion
    environment:
      name: pypi
      url: https://pypi.org/p/<project-name>
    permissions:
```

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:2]: If it's not against some policy, let's link https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/ — we're trying to keep it up-to-date over in PyPUG, which often happens sooner than GH gets docs updated.

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:43]: explanatory comment plz?

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:53]: This might be a good place to give out a hardening recommendation:
```suggestion
        uses: pypa/gh-action-pypi-publish@release/v1  # update `release/v1` to a commit SHA if you use Dependabot
```

**Comment by webknjaz** [comment]
@woodruffw love that this is moving somewhere! Thank you for getting to this sooner than me :) I've been frustrated with how many people get pre-historic workflows by default and don't even know it...

One extra thing to consider — it might be useful to also stick Sigstore signing right into the starter. OTOH, giving people a link to the guide might be an alternative.

File: ci/python-publish.yml
Comment (Plentywolf) [reply ci/python-publish.yml:52]: Master script??

Comment (Plentywolf) [reply ci/python-publish.yml:52]: 52 yes

**Comment by webknjaz** [comment]
> I've been in discussion with @woodruffw, @di and the PyPI team about this.

Linking some context: https://github.com/pypa/gh-action-pypi-publish/pull/123#issuecomment-1982201999

File: ci/python-publish.yml
Comment (woodruffw) [reply ci/python-publish.yml:15]: Makes sense, I'll re-add!

File: ci/python-publish.yml
Comment (woodruffw) [reply ci/python-publish.yml:2]: Makes sense to me. I'll defer to GH folks for any actual policy here 🙂

File: ci/python-publish.yml
Comment (woodruffw) [reply ci/python-publish.yml:30]: I'm -0 on this, for three reasons:

1. Fewer users know about `pipx`, so putting it here increases (very, very slightly) the cognitive complexity of the sample workflow. I recognize that this is a self-defeating argument, however, since this is the kind of thing needed to raise user awareness of `pipx` 🙂 
2. IMO `pipx run --spec` is a lot of indirection to introduce here: it requires users to think about `pipx`, about a venv for `build`, about a subprocess spawning `python` in that venv, etc. It's simpler for people like us, but I think this workflow also needs to cater to first-time publishers.
3. `pipx`'s availability on different runner images seems mixed/not subject to automated checks: https://github.com/actions/runner-images/issues/7590. Most users probably won't hit this since they'll be on `ubuntu-latest`, but it's another edge case we can dodge.

File: ci/python-publish.yml
Comment (woodruffw) [reply ci/python-publish.yml:42]: Agreed, although this will produce an incongruity between this workflow and the other GitHub docs. I suppose we should update all of them to suggest using an environment.

File: ci/python-publish.yml
Comment (woodruffw) [reply ci/python-publish.yml:53]: No objection, although I'm not sure if there's precedent in the starter workflows for this. I'll look.

File: ci/python-publish.yml
Comment (woodruffw) [reply ci/python-publish.yml:52]: This might be a dialect thing: I've only ever heard "PyPI" without the *the* definite article, whereas expanded it would be "*the* Python Package Index" 🙂

File: ci/python-publish.yml
Comment (di) [reply ci/python-publish.yml:30]: I agree.

**Comment by woodruffw** [comment]
> One extra thing to consider — it might be useful to also stick Sigstore signing right into the starter. OTOH, giving people a link to the guide might be an alternative.

I'd personally like to shy away from suggesting the Sigstore action here for now, if only because (with PEP 740) it'll become obsolete and integrated directly into the publishing flow 🙂

File: ci/python-publish.yml
Comment (webknjaz) [reply ci/python-publish.yml:42]: All the docs and examples I could reach already use environments called `pypi` and `testpypi` by default. Not `release` since that name is ambiguous and doesn't represent a specific deployment target (the PyPI).

Comment (webknjaz) [reply ci/python-publish.yml:42]: GH docs you contributed earlier seem to use the `release` environment (so does the placeholder in Warehouse, which I dislike for the same reason).

File: ci/python-publish.yml
Comment (webknjaz) [reply ci/python-publish.yml:30]: Okay, I withdraw this idea.

File: ci/python-publish.yml
Comment (webknjaz) [reply ci/python-publish.yml:57]: My understanding is that they usually demand commit SHA in all starter workflows which is what got us into this mess in the first place. So I figured this might be a good middle ground 🤷‍♂️

Comment (woodruffw) [reply ci/python-publish.yml:57]: Yeah, looks like `gem-push.yml` has a commit SHA, and a comment suggesting a change to a symbolic ref for automatic changes. I'll do something similar here 🙂

Comment (woodruffw) [reply ci/python-publish.yml]: Did a version of what `gem-push.yml` does, PTAL!

File: ci/python-publish.yml
Comment (woodruffw) [reply ci/python-publish.yml:45]: I was thinking of [this example](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-pypi#updating-your-github-actions-workflow), which (regrettably) doesn't use an environment. I'll add an environment here and work on updating that doc separately.

Comment (woodruffw) [reply ci/python-publish.yml:45]: Added an environment example with some comments, PTAL!

File: ci/python-publish.yml
Comment (webknjaz) [reply ci/python-publish.yml:63]: It's a grammar thing (think: *the* US). Though I indeed saw various usage patterns as well. OTOH PyPI is unique enough not to need an article (people don't tend to confuse it with Test PyPI).

I'm likely confusing it with the preferred usage of "the PyPA". I did some googling and only found it the long form: https://packaging.python.org/en/latest/glossary/#term-pypi.org. The FAQ on pypi.org doesn't use the article in front of the abbreviation.

https://english.stackexchange.com/a/82827/68752 suggests this depends on how one would pronounce an abbreviation.

Anyway, I'm going to trust your instinct. Please, disregard this suggestion.

File: ci/python-publish.yml
Comment (webknjaz) [reply ci/python-publish.yml]: Personally, I prefer the rolling branch but I don't think the GH folks would agree to that.. Dependabot wouldn't update it, though (until converted to a tag or a commit hash).
I suppose your variant will work for most.

File: ci/python-publish.yml
Comment (webknjaz) [reply ci/python-publish.yml:45]: Looks good. Normally I like making versioned URLs. But that might be confusing to newbies. Although, if there's an easily retrieveable version in the release event, sticking it into the example comment would be a nice touch.

Comment (woodruffw) [reply ci/python-publish.yml:45]: I think there is, although there's no guarantee that the release version is a match for the PyPI version (user error, but also because of PEP 440 normalization/rewrites). So I'm inclined to leave this unversioned for now, although I don't have a super strong opinion here 🙂

Comment (webknjaz) [reply ci/python-publish.yml:45]: I think it's fine if it's not normalized and the URL might be off (although, personally I started embedding `workflow_dispatch` inputs validation for PEP 440), and if it's in a comment, as a hint. The most reliable way of getting that would be unpacking a wheel or sdist and grabbing it from the structured metadata since the packaging machinery usually normalizes it already. I won't insist, though.

**Comment by woodruffw** [comment]
This should be good to go. I think it might make sense to review and land this before https://github.com/github/docs/pull/32146, since the workflow changes here will need to be reflected there as well.

**Comment by woodruffw** [comment]
Gentle ping for review here!

(@jhutchings1 I'm calling in that promise 😉)

**Comment by woodruffw** [comment]
Another gentle ping for review here: I believe this should be merged before https://github.com/github/docs/pull/32146 goes forwards, since those docs will link to this workflow.

**Comment by jhutchings1** [comment]
@nebuk89 Can you help get this prioritized?

File: ci/python-publish.yml
Comment (pandafy) [inline ci/python-publish.yml:66]: There is a newer release of the action available. 

```suggestion
        uses: pypa/gh-action-pypi-publish@ec4db0b4ddc65acdf4bff5fa45ac92d78b56bdf0 # v1.9.0
        
```

**Comment by woodruffw** [comment]
Gentle ping for review here -- it'd be _really_ nice to get this landed, since other parts of GitHub's docs now refer extensively to PyPI's support for OIDC/Trusted Publishing.

**Comment by pandafy** [comment]
It took me sometime to debug why trusted publishing on my package was not working, and it was due to the old commit hash present in the starter code. Getting this merged would be wonderful.

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:54]: @woodruffw how about mentioning that versioned URL as optional?
```suggestion
      # url: https://pypi.org/p/YOURPROJECT
      #
      # ALTERNATIVE: if your GitHub Release name is the PyPI project version string
      # ALTERNATIVE: exactly, uncomment the following line instead:
      # url: https://pypi.org/project/YOURPROJECT/${{ github.event.release.name }}
```

**Comment by jhutchings1** [comment]
@actions/actions-workflow-development-reviewers Can we get a PR review on this? This is the best practice from the ecosystem maintainers and this PR has been sitting for a very long time. cc: @nebuk89 @chrispat @steiza

**Comment by webknjaz** [comment]
@woodruffw perhaps, we could stick attestations into this, finally?

**Comment by woodruffw** [comment]
> @woodruffw perhaps, we could stick attestations into this, finally?

Do we need to do anything explicitly for them here? The way we've designed the attestation support in `gh-action-pypi-publish` means that they'll become the default once we switch the flag, so (IIUC) we should be able to leave this workflow mostly as-is and users will begin attesting as they upgrade to the version that makes it the default.

That being said, it probably makes sense to update the version/commit ref in this PR once again. I'll do that now.

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:70]:
```suggestion
        uses: pypa/gh-action-pypi-publish@fb13cb306901256ace3dab689990e13a5550ffaa  # v1.11.0
```

**Comment by woodruffw** [comment]
Gentle ping -- this has been sitting for a while now and is now behind the other GitHub docs (and PyPI/PyPUG docs) in terms of what it contains.

CC @jhutchings1 @nebuk89

**Comment by jhutchings1** [comment]
I'm really bummed this one hasn't gotten traction. I'm not an employee anymore, so don't have any official seat, but @chrispat  anything you can do to get this one some love?

**Comment by jhutchings1** [comment]
I'm really bummed this one hasn't gotten traction. I'm not an employee anymore, so don't have any official seat, but @chrispat  anything you can do to get this one some love?

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:70]:
```suggestion
        uses: pypa/gh-action-pypi-publish@1f5d4ec244f65dce93685ee3e98e77123f090866  # v1.12.1
```

File: ci/python-publish.yml
Comment (webknjaz) [inline ci/python-publish.yml:70]:
```suggestion
        uses: pypa/gh-action-pypi-publish@15c56dba361d8335944d31a2ecd17d700fc7bcbc  # v1.12.2
```

File: ci/python-publish.yml
Comment (steiza) [reply ci/python-publish.yml:70]: Separately: in my test case I had a custom directory. This worked fine, after I found `packages-dir`. Even though it's the default, should we provide `dist/` here like we do in the above commands?

I could go either way.

```suggestion
        uses: pypa/gh-action-pypi-publish@15c56dba361d8335944d31a2ecd17d700fc7bcbc  # v1.12.2
        with:
          packages-dir: dist/
```

**Review by steiza — APPROVED** [review]
3 minor comments, otherwise looks good to me! I was able to use this workflow successfully.

File: ci/python-publish.yml
Comment (woodruffw) [reply ci/python-publish.yml:70]: +1 to making it explicit and pinning on `release/v1` instead of a digest -- normally I'd say we should suggest the digest, but the long review turnaround here indicates that doing so will be a maintenance/staleness risk 🙂

File: ci/python-publish.yml
Comment (woodruffw) [inline ci/python-publish.yml:70]:
```suggestion
        uses: pypa/gh-action-pypi-publish@release/v1
```

**Comment by woodruffw** [comment]
Thanks @steiza, applied!

**Review by steiza — APPROVED** [review]
Fantastic; thanks! I think this is ready to merge.

**Comment by woodruffw** [comment]
Thanks a ton @steiza and @elbrenn!

**Comment by webknjaz** [comment]
Thanks @elbrenn and @steiza!

**Pull request merged**
